homeassistant/components/sensor/whois.py

```diff
@@ -0,0 +1,135 @@
+    vol.Required(CONF_DOMAIN): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string
+})
+
+
+def setup_platform(hass, config, add_devices, disovery_info=None):
+    """Set up the WHOIS sensor."""
+    from pythonwhois import get_whois
+    from pythonwhois.shared import WhoisException
+
+    domain = config.get(CONF_DOMAIN)
+    name = config.get(CONF_NAME)
+
+    try:
+        if 'expiration_date' in get_whois(domain, normalized=True):
+            add_devices([WhoisSensor(name, domain)], True)
+        else:
+            _LOGGER.warning("Failed to perform WHOIS lookup for %s",
+                            domain)
+            return False
+    except WhoisException as ex:
+        _LOGGER.error("Exception %s occurred during WHOIS lookup for %s",
+                      ex,
+                      domain)
+        return False
+
+
+class WhoisSensor(Entity):
+    """Implementation of a WHOIS sensor."""
+
+    def __init__(self, name, domain):
+        """Initialize the sensor."""
+        from pythonwhois import get_whois
+
+        self.whois = get_whois
+
+        self._name = name
+        self._domain = domain
+
+        self._state = None
+        self._data = None
+        self._updated_date = None
+        self._expiration_date = None
+        self._name_servers = []
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """The icon to represent this sensor."""
+        return 'mdi:calendar-clock'
+
+    @property
+    def unit_of_measurement(self):
+        """The unit of measurement to present the value in."""
+        return 'days'
+
+    @property
+    def state(self):
+        """Return the expiration days for hostname."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Get the more info attributes."""
+        if self._data:
+            updated_formatted = self._updated_date.isoformat()
+            expires_formatted = self._expiration_date.isoformat()
+
+            return {
+                ATTR_NAME_SERVERS: ' '.join(self._name_servers),
+                ATTR_REGISTRAR: self._data['registrar'][0],
+                ATTR_UPDATED: updated_formatted,
+                ATTR_EXPIRES: expires_formatted,
+            }
+
+    def update(self):
+        """Get the current WHOIS data for hostname."""
+        from pythonwhois.shared import WhoisException
+
+        try:
+            response = self.whois(self._domain, normalized=True)
+        except WhoisException as ex:
+            _LOGGER.error("Exception %s occurred during WHOIS lookup", ex)
+            return False
+
+        if response:
+            self._data = response
+
+            if self._data['nameservers']:
+                self._name_servers = self._data['nameservers']
+
+            self._expiration_date = self._data['expiration_date'][0]
+            self._updated_date = self._data['updated_date'][0]
```

Comment:
With `home-assistant.io` I get:

```bash
  File "/home/fab/Documents/repos/home-assistant/homeassistant/components/sensor/whois.py", line 131, in update
    self._updated_date = self._data['updated_date'][0]
KeyError: 'updated_date'
```

Reply:
Seems my assumption of its existence was incorrect, I've made it conditional now. Resolved the other comments too.